core/base/mergeTreePrincipalGeodesics/MergeTreePrincipalGeodesicsBase.h

```diff
@@ -0,0 +1,893 @@
+      std::vector<std::vector<std::tuple<ftm::idNode, ftm::idNode, double>>>
+        &matchings,
+      bool transposeVector = true) {
+      reconstructionErrors.resize(inputTrees.size());
+      matchings.resize(inputTrees.size());
+      dataType reconstructionError = 0.0;
+#ifdef TTK_ENABLE_OPENMP
+#pragma omp parallel for schedule(dynamic) num_threads(this->threadNumber_) reduction(+:reconstructionError)
+#endif
+      for(unsigned int i = 0; i < inputTrees.size(); ++i) {
+        ftm::MergeTree<dataType> reconstructedTree;
+        getMultiInterpolation<dataType>(barycenter, vS, v2s, vSize,
+                                        allTreesTs[i], reconstructedTree,
+                                        transposeVector);
+        dataType error;
+        computeOneDistance<dataType>(
+          reconstructedTree, inputTrees[i], matchings[i], error, true);
+        reconstructionError += error / inputTrees.size();
+        reconstructionErrors[i] = error;
+      }
+      return reconstructionError;
+    }
+
+    template <class dataType>
+    dataType computeReconstructionError(
+      ftm::MergeTree<dataType> &barycenter,
+      std::vector<ftm::MergeTree<dataType>> &inputTrees,
+      std::vector<std::vector<std::vector<double>>> &vS,
+      std::vector<std::vector<std::vector<double>>> &v2s,
+      std::vector<std::vector<double>> &allTreesTs,
+      std::vector<double> &reconstructionErrors) {
+      std::vector<std::vector<double *>> pVS, pV2s;
+      vectorOfVectorsToPointers(vS, pVS);
+      vectorOfVectorsToPointers(v2s, pV2s);
+      std::vector<std::vector<std::tuple<ftm::idNode, ftm::idNode, double>>>
+        matchings;
+      return computeReconstructionError(barycenter, inputTrees, pVS, pV2s,
+                                        vS[0].size(), allTreesTs,
+                                        reconstructionErrors, matchings, false);
+    }
+
+    template <class dataType>
+    dataType computeReconstructionError(
+      ftm::MergeTree<dataType> &barycenter,
+      std::vector<ftm::MergeTree<dataType>> &inputTrees,
+      std::vector<std::vector<std::vector<double>>> &vS,
+      std::vector<std::vector<std::vector<double>>> &v2s,
+      std::vector<std::vector<double>> &allTreesTs) {
+      std::vector<double> reconstructionErrors;
+      return computeReconstructionError(
+        barycenter, inputTrees, vS, v2s, allTreesTs, reconstructionErrors);
+    }
+
+    //----------------------------------------------------------------------------
+    // Interpolation
+    //----------------------------------------------------------------------------
+    double getGeodesicVectorMiddle(std::vector<double> &v,
+                                   std::vector<double> &v2) {
+      std::vector<double> vProj, v2Proj;
+      ttk::Geometry::addVectorsProjection(v, v2, vProj, v2Proj);
+
+      double alpha = 0.0;
+      int cptDivide = 0;
+      for(unsigned int i = 0; i < vProj.size(); ++i) {
+        if(std::abs(v2Proj[i]) < 1e-12)
```

Comment:
same remark as before regarding `1e-12`

Reply:
Done in the next commit!

Reply:
Done in the next commit!
(with almost the same comment, since the power is fixed)